libs/server/Storage/Session/MainStore/MainStoreOps.cs

```diff
@@ -558,16 +556,39 @@ public unsafe GarnetStatus RENAME(ArgSlice oldKeySlice, ArgSlice newKeySlice, St
                         var ptrVal = (byte*)memoryHandle.Pointer;
 
                         RespReadUtils.ReadUnsignedLengthHeader(out var headerLength, ref ptrVal, ptrVal + o.Length);
-                        var value = SpanByte.FromPinnedPointer(ptrVal, headerLength);
-                        SET(ref newKey, ref value, ref context);
-
-                        memoryHandle.Dispose();
-                        o.Memory.Dispose();
-
-                        // Delete the old key
-                        DELETE(ref oldKey, StoreType.Main, ref context, ref objectContext);
 
-                        returnStatus = GarnetStatus.OK;
+                        // Find expiration time of the old key
+                        var expireSpan = new SpanByteAndMemory();
+                        var ttlStatus = TTL(ref oldKey, storeType, ref expireSpan, ref context, ref objectContext, true);
+
+                        if (ttlStatus == GarnetStatus.OK && !expireSpan.IsSpanByte)
+                        {
+                            var expireMemoryHandle = expireSpan.Memory.Memory.Pin();
+                            var expirePtrVal = (byte*)expireMemoryHandle.Pointer;
+                            RespReadUtils.TryRead64Int(out var expireTimeMs, ref expirePtrVal, expirePtrVal + expireSpan.Length, out var _);
+
+                            // If the key has an expiration, set the new key with the expiration
+                            if (expireTimeMs > 0)
+                            {
+                                SETEX(newKeySlice, new ArgSlice(ptrVal, headerLength), TimeSpan.FromMilliseconds(expireTimeMs), ref context);
+                            }
+                            else if (expireTimeMs == -1) // Its possible to have expire as 0 or -2, in those cases we don't SET the new key
```

Comment:
Check what expire time of 0 and -2 indicate.

Reply:
To my understanding:
0 - Key expired
-1 - No expiry associated with key
-2 - Key does not exist

Reply:
@TalZaccai Is right. That is the reason I added a comment saying we shouldn't call `SET` in those cases because it could create a KEY that is not there now (or shouldn't be there now). I will add the reason behind it in the comments to clarify.